File: webapp/src/types/playbook.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Select from "react-select/src/Select";

export interface Playbook {
    id?: string;
    title: string;
    description: string;
    team_id: string;
    create_public_incident: boolean;
    checklists: Checklist[];
    propertylist: Propertylist;
    member_ids: string[];
    broadcast_channel_id: string;
    reminder_message_template: string;
    reminder_timer_default_seconds: number;
}

export interface PlaybookNoChecklist {
    id?: string;
    title: string;
    description: string;
    team_id: string;
    create_public_incident: boolean;
    num_stages: number;
    num_steps: number;
    member_ids: string[];
}

export interface FetchPlaybooksNoChecklistReturn {
    total_count: number;
    page_count: number;
    has_more: boolean;
    items: PlaybookNoChecklist[];
}

export interface FetchIncidentsParams {
    sort?: string;
    direction?: string;
}

export interface Propertylist {
    id?: string;
    title: string;
    items: PropertylistItem[];
}
export interface Checklist {
    title: string;
    items: ChecklistItem[];
}

export enum ChecklistItemState {
    Open = '',
    InProgress = 'in_progress',
    Closed = 'closed',
}

export interface PropertylistItem {
    id?: string;
    title: string
    type: PropertyType
    is_mandatory: boolean
    selection?: Selectionlist
    freetext?: TextOption
}

export interface TextOption {
    value: string
    badge_style?: BadgeStyle
}

export interface Selectionlist {
    items: SelectionlistItem[]
    is_multiselect: boolean;
    selected_id: string;
}

export interface SelectionlistItem {
    id?: string;
    value: string;
    badge_style?: BadgeStyle
}

export interface BadgeStyle {
    badge_color: string;
    text_color: string;
}

export enum PropertyType {
    Freetext = 'Freetext',
    Selection = 'Selection',
}

export interface ChecklistItem {
    title: string;
    description: string;
    state: ChecklistItemState;
    state_modified?: number;
    state_modified_post_id?: string;
    assignee_id?: string;
    assignee_modified?: number;
    assignee_modified_post_id?: string;
    command: string;
    command_last_run: number;
}

export function emptyPlaybook(): Playbook {
    return {
        id: '',
        title: '',
        description: '',
        team_id: '',
        create_public_incident: false,
        checklists: [emptyChecklist()],
        propertylist: emptyPropertylist(),
        member_ids: [],
        broadcast_channel_id: '',
        reminder_message_template: '',
        reminder_timer_default_seconds: 0,
    };
}

export function emptyChecklist(): Checklist {
    return {
        title: 'Default Checklist',
        items: [emptyChecklistItem()],
    };
}

export function emptyPropertylist(): Propertylist {
    return {
        id:'',
        title: 'Default Checklist',
        items: [emptyPropertylistItem()],
    };
}

export function emptyPropertylistItem(): PropertylistItem {
    return {
        id: '',
        title: '',
        type: PropertyType.Freetext,
        is_mandatory: false,
        selection: emptySelectionlist(),
        freetext: emptyFreetextOption()
    };
}

export function emptySelectionlist(): Selectionlist {
    return {
        items:  [emptySelectionlistItem()],
        selected_id: '',
        is_multiselect: false,
    };
}

export function emptySelectionlistItem(): SelectionlistItem {
    return {
        id: '',
        value: '',
        badge_style: emptyBadgeStyle()
    };
}

export function emptyFreetextOption(): TextOption {
    return {
    value: '',
    badge_style: emptyBadgeStyle()
    };
}

export function emptyBadgeStyle(): BadgeStyle {
    return {
        badge_color: '',
        text_color: '',
    };
}

export function emptyChecklistItem(): ChecklistItem {
    return {
        title: '',
        state: ChecklistItemState.Open,
        command: '',
        description: '',
        command_last_run: 0,
    };
}

export const newChecklistItem = (title = '', description = '', command = '', state = ChecklistItemState.Open): ChecklistItem => ({
    title,
    description,
    command,
    command_last_run: 0,
    state,
});

export const newPropertylist = (id = '', title = '', items = [newPropertylistItem()] ): Propertylist => ({
    id,
    title,
    items
});

export const newPropertylistItem = (id = '', title = '', optional = false, type = PropertyType.Freetext, selection = newSelectionlist(), freetext = newFreetextOption()): PropertylistItem => ({
    id,
    title,
    is_mandatory: optional,
    type,
    selection,
    freetext,
});

export const newSelectionlist = (items = [newSelectionlistItem()], selected_id = '', is_multiselect = false): Selectionlist => ({
    items,
    selected_id,
    is_multiselect
});

export const newSelectionlistItem = (id = '', value = '', badge_style = newBadgeStyle()): SelectionlistItem => ({
    id,
    value,
    badge_style
});

export const newBadgeStyle = (badge_color = '', text_color = ''): BadgeStyle => ({
    badge_color,
    text_color,
});

export interface TextOption {
    value: string
    badge_style?: BadgeStyle
}

export const newFreetextOption = (value='', badge_style = newBadgeStyle()): TextOption => ({
    value,
    badge_style
});

// eslint-disable-next-line
export function isPlaybook(arg: any): arg is Playbook {
    return arg &&
        typeof arg.id === 'string' &&
        typeof arg.title === 'string' &&
        typeof arg.team_id === 'string' &&
        typeof arg.create_public_incident === 'boolean' &&
        arg.checklists && Array.isArray(arg.checklists) && arg.checklists.every(isChecklist) &&
        arg.properties && Array.isArray(arg.properties) && arg.checklists.every(isPropertyItem) &&
        arg.member_ids && Array.isArray(arg.member_ids) && arg.checklists.every((id: any) => typeof id === 'string') &&
        typeof arg.broadcast_channel_id === 'string';
}

// eslint-disable-next-line
export function isChecklist(arg: any): arg is Checklist {
    return arg &&
        typeof arg.title === 'string' &&
        arg.items && Array.isArray(arg.items) && arg.items.every(isChecklistItem);
}

// eslint-disable-next-line
export function isChecklistItem(arg: any): arg is ChecklistItem {
    return arg &&
        typeof arg.title === 'string' &&
        typeof arg.state_modified === 'number' &&
        typeof arg.state_modified_post_id === 'string' &&
        typeof arg.assignee_id === 'string' &&
        typeof arg.assignee_modified === 'number' &&
        typeof arg.assignee_modified_post_id === 'string' &&
        typeof arg.state === 'string' &&
        typeof arg.command === 'string' &&
        typeof arg.command_last_run === 'number';
}

// eslint-disable-next-line
export function isPropertyItem(arg: any): arg is PropertylistItem {
    return arg &&
        typeof arg.title === 'string' &&
        typeof arg.state_modified === 'number' &&
        typeof arg.state_modified_post_id === 'string' &&
        typeof arg.assignee_id === 'string' &&
        typeof arg.assignee_modified === 'number' &&
        typeof arg.assignee_modified_post_id === 'string' &&
        typeof arg.state === 'string' &&
        typeof arg.command === 'string' &&
        typeof arg.command_last_run === 'number';
}

File: server/playbook/playbook.go
package playbook

import (
	"encoding/json"

	"github.com/pkg/errors"
)

// ErrNotFound used to indicate entity not found.
var ErrNotFound = errors.New("not found")

// Playbook represents the planning before an incident type is initiated.
type Playbook struct {
	ID                          string       `json:"id"`
	Title                       string       `json:"title"`
	Description                 string       `json:"description"`
	TeamID                      string       `json:"team_id"`
	CreatePublicIncident        bool         `json:"create_public_incident"`
	CreateAt                    int64        `json:"create_at"`
	DeleteAt                    int64        `json:"delete_at"`
	NumStages                   int64        `json:"num_stages"`
	NumSteps                    int64        `json:"num_steps"`
	Checklists                  []Checklist  `json:"checklists"`
	Propertylist                Propertylist `json:"propertylist"`
	MemberIDs                   []string     `json:"member_ids"`
	BroadcastChannelID          string       `json:"broadcast_channel_id"`
	ReminderMessageTemplate     string       `json:"reminder_message_template"`
	ReminderTimerDefaultSeconds int64        `json:"reminder_timer_default_seconds"`
}

func (p Playbook) Clone() Playbook {
	newPlaybook := p
	var newChecklists []Checklist
	for _, c := range p.Checklists {
		newChecklists = append(newChecklists, c.Clone())
	}

	var newPropertylist Propertylist
	newPropertylist = p.Propertylist.Clone()

	newPlaybook.Propertylist = newPropertylist
	newPlaybook.Checklists = newChecklists
	newPlaybook.MemberIDs = append([]string(nil), p.MemberIDs...)
	return newPlaybook
}

func (p Playbook) MarshalJSON() ([]byte, error) {
	type Alias Playbook

	old := Alias(p.Clone())
	// replace nils with empty slices for the frontend
	if old.Checklists == nil {
		old.Checklists = []Checklist{}
	}
	for j, cl := range old.Checklists {
		if cl.Items == nil {
			old.Checklists[j].Items = []ChecklistItem{}
		}
	}

	if old.Propertylist.Items == nil {
		old.Propertylist.Items = []PropertylistItem{}
	}
	if old.MemberIDs == nil {
		old.MemberIDs = []string{}
	}

	return json.Marshal(old)
}

// Propertylist represents collection of properties in a playbook
type Propertylist struct {
	ID    string             `json:"id"`
	Title string             `json:"title"`
	Items []PropertylistItem `json:"items"`
}

// Checklist represents a checklist in a playbook
type Checklist struct {
	ID    string          `json:"id"`
	Title string          `json:"title"`
	Items []ChecklistItem `json:"items"`
}

func (c Checklist) Clone() Checklist {
	newChecklist := c
	newChecklist.Items = append([]ChecklistItem(nil), c.Items...)
	return newChecklist
}

func (c Propertylist) Clone() Propertylist {
	newPropertylist := c
	return newPropertylist
}

func (c PropertylistItem) Clone() PropertylistItem {
	newPropertyItem := c
	return newPropertyItem
}

func (c TextOption) Clone() TextOption {
	newTextOption := c
	return newTextOption
}

func (c Selectionlist) Clone() Selectionlist {
	newSelectionlist := c
	return newSelectionlist
}

func (c SelectionlistItem) Clone() SelectionlistItem {
	newSelectionlistItem := c
	return newSelectionlistItem
}

// ChecklistItem represents an item in a checklist
type ChecklistItem struct {
	ID                     string `json:"id"`
	Title                  string `json:"title"`
	State                  string `json:"state"`
	StateModified          int64  `json:"state_modified"`
	StateModifiedPostID    string `json:"state_modified_post_id"`
	AssigneeID             string `json:"assignee_id"`
	AssigneeModified       int64  `json:"assignee_modified"`
	AssigneeModifiedPostID string `json:"assignee_modified_post_id"`
	Command                string `json:"command"`
	CommandLastRun         int64  `json:"command_last_run"`
	Description            string `json:"description"`
}

type GetPlaybooksResults struct {
	TotalCount int        `json:"total_count"`
	PageCount  int        `json:"page_count"`
	HasMore    bool       `json:"has_more"`
	Items      []Playbook `json:"items"`
}

// PropertylistItem represents an item in a checklist
type PropertylistItem struct {
	ID          string        `json:"id"`
	Title       string        `json:"title"`
	Type        string        `json:"type"`
	IsMandatory bool          `json:"is_mandatory"`
	Selection   Selectionlist `json:"selection"`
	Treetext    TextOption    `json:"freetext"`
}

type TextOption struct {
	Value      string     `json:"value"`
	BadgeStyle BadgeStyle `json:"badge_style"`
}

type Selectionlist struct {
	Items         []SelectionlistItem `json:"items"`
	IsMultiselect bool                `json:"is_multiselect"`
	SelectedId    string              `json:"selected_id"`
}

type SelectionlistItem struct {
	ID         string     `json:"id"`
	Value      string     `json:"value"`
	BadgeStyle BadgeStyle `json:"badge_style"`
}

type BadgeStyle struct {
	BadgeColor string `json:"badge_color"`
	TextColor  string `json:"text_color"`
}

// MarshalJSON customizes the JSON marshalling for GetPlaybooksResults by rendering a nil Items as
// an empty slice instead.
func (r GetPlaybooksResults) MarshalJSON() ([]byte, error) {
	type Alias GetPlaybooksResults

	if r.Items == nil {
		r.Items = []Playbook{}
	}

	aux := &struct {
		*Alias
	}{
		Alias: (*Alias)(&r),
	}

	return json.Marshal(aux)
}

// RequesterInfo holds the userID and permissions for the user making the request
type RequesterInfo struct {
	UserID          string
	TeamID          string
	UserIDtoIsAdmin map[string]bool

	// MemberOnly filters playbooks to those for which UserId is a member
	MemberOnly bool
}

// Service is the playbook service for managing playbooks
// userID is the user initiating the event.
type Service interface {
	// Get retrieves a playbook. Returns ErrNotFound if not found.
	Get(id string) (Playbook, error)
	// Create creates a new playbook
	Create(playbook Playbook, userID string) (string, error)
	// GetPlaybooks retrieves all playbooks
	GetPlaybooks() ([]Playbook, error)
	// GetPlaybooksForTeam retrieves all playbooks on the specified team given the provided options
	GetPlaybooksForTeam(requesterInfo RequesterInfo, teamID string, opts Options) (GetPlaybooksResults, error)
	// Update updates a playbook
	Update(playbook Playbook, userID string) error
	// Delete deletes a playbook
	Delete(playbook Playbook, userID string) error
}

// Store is an interface for storing playbooks
type Store interface {
	// Get retrieves a playbook
	Get(id string) (Playbook, error)
	// Create creates a new playbook
	Create(playbook Playbook) (string, error)
	// GetPlaybooks retrieves all playbooks
	GetPlaybooks() ([]Playbook, error)
	// GetPlaybooksForTeam retrieves all playbooks on the specified team
	GetPlaybooksForTeam(requesterInfo RequesterInfo, teamID string, opts Options) (GetPlaybooksResults, error)
	// Update updates a playbook
	Update(playbook Playbook) error
	// Delete deletes a playbook
	Delete(id string) error
}

// Telemetry defines the methods that the Playbook service needs from the RudderTelemetry.
// userID is the user initiating the event.
type Telemetry interface {
	// CreatePlaybook tracks the creation of a playbook.
	CreatePlaybook(playbook Playbook, userID string)

	// UpdatePlaybook tracks the update of a playbook.
	UpdatePlaybook(playbook Playbook, userID string)

	// DeletePlaybook tracks the deletion of a playbook.
	DeletePlaybook(playbook Playbook, userID string)
}

const (
	ChecklistItemStateOpen       = ""
	ChecklistItemStateInProgress = "in_progress"
	ChecklistItemStateClosed     = "closed"
)

func IsValidChecklistItemState(state string) bool {
	return state == ChecklistItemStateClosed ||
		state == ChecklistItemStateInProgress ||
		state == ChecklistItemStateOpen
}

func IsValidChecklistItemIndex(checklists []Checklist, checklistNum, itemNum int) bool {
	return checklists != nil && checklistNum >= 0 && itemNum >= 0 && checklistNum < len(checklists) && itemNum < len(checklists[checklistNum].Items)
}
